### ISSUE TITLE
class PubSubClient' has no member named 'state

### ISSUE THREAD

**Issue by Santosh1971** [internetofhomethings/ESP8266-MQTT-HTTP-Server#1]
I added
void \* pvPortZalloc(int size); as you suggested last time in another program.

I am getting following error while compiling with Arduino 1.6.5

webandmqtt_server.ino: In function 'void MqttServer_Init()':
webandmqtt_server:1084: error: 'class PubSubClient' has no member named 'setServer'
webandmqtt_server:1085: error: 'class PubSubClient' has no member named 'setCallback'
'class PubSubClient' has no member named 'state'

Serial.print(client.state());

pl. suggest the solution.
Also can I check this with different MQTT broker?
In this example have you setup broker locally? 

regards,
Santosh

**Comment by internetofhomethings** [comment]
I had installed the PubSubClient library in the Arduino IDE per this post: http://internetofhomethings.com/homethings/?p=1121  

See the section "Arduino IDE Sketch"

You should not get the compile errors noted if the PubSubClient was installed in that manner as it comes directly from the source and does include the members named in the errors.

Sure, you can use any MQTT broker you wish. In this example, near the top of the sketch, just change the following 3 lines to point to your broker:

#define mqtt_server "test.mosquitto.org"
#define mqtt_user "your_username"
#define mqtt_password "your_password"

In this example, I am using the broker "test.mosquitto.org". It is not local. I chose that one because it is free and does not require a username/password. That is great for testing but obviously a un/pw should be added for security for a serious application.

**Comment by Santosh1971** [comment]
True I just updated the PubSubClient and now it is compiling.
Sorry for bothering you.
and Thanks for clarifying!
Thanks.

**Comment by Santosh1971** [comment]
Dave ,one issue in testing html page

When I send the MQTT request for LED on and off,it works fine, and I get a Reply,like LED is ON/OFF, but for GetSensors I am not getting any response.
However on serial port I am getting this data

Message arrived [MyMqttSvrRqst] {
"Ain0":"53.00",
"Ain1":"52.00",
"Ain2":"52.00",
"Ain3":"50.00",
"Ain4":"103.00",
"Ain5":"47.00",
"Ain6":"184.00",
"Ain7":"56.00",
"SYS_Heap":"22080",
"SYS_Time":"1672"
}

Pl. suggest!

**Comment by Santosh1971** [comment]
with HTTP everything is working fine.
I get this response
{ "Ain0":"74.00", "Ain1":"65.00", "Ain2":"62.00", "Ain3":"62.00", "Ain4":"70.00", "Ain5":"188.00", "Ain6":"198.00", "Ain7":"206.00", "SYS_Heap":"27864", "SYS_Time":"1973" }

**Comment by Santosh1971** [comment]
when I commented out few of the following message to reduce the MQTT reply message size..

 case Get_SENSORS:
            //Create JSON return string
            if(servertype == SVR_HTTP_LIB) {
                s += "Content-Type: application/json\r\n\r\n";
            }
            jsonEncode(FIRSTJSON,&s,"Ain0", Ain0);
            jsonEncode(NEXTJSON,&s,"Ain1", Ain1);
            jsonEncode(NEXTJSON,&s,"Ain2", Ain2);
            jsonEncode(NEXTJSON,&s,"Ain3", Ain3);
          //  jsonEncode(NEXTJSON,&s,"Ain4", Ain4);
         //   jsonEncode(NEXTJSON,&s,"Ain5", Ain5);
         //   jsonEncode(NEXTJSON,&s,"Ain6", Ain6);
         //   jsonEncode(NEXTJSON,&s,"Ain7", Ain7);
            v = system_get_free_heap_size();
          //  jsonEncode(NEXTJSON,&s,"SYS_Heap", v);
            v = millis()/1000;
            jsonEncode(LASTJSON,&s,"SYS_Time", v);
            break;
        default:
            break;

It started working.
Is it the reason?
pl. advise.
Regards,
Santosh

**Comment by Santosh1971** [comment]
in case of MQTT request I get following response
{ "Ain0":"201.00", "Ain1":"175.00", "Ain2":"198.00", "Ain3":"198.00", "SYS_Time":"1764" }

In case of HTTP request I get
{ "Ain0":"192.00", "Ain1":"199.00", "Ain2":"172.00", "Ain3":"198.00", "Ain4":"undefined", "Ain5":"undefined", "Ain6":"undefined", "Ain7":"undefined", "SYS_Heap":"undefined", "SYS_Time":"1845" }

How come Ain6: undefined coming.. ?
Pl. advise
regards,
Santosh

**Comment by Santosh1971** [comment]
One more expectation:-
As both HTTP request and MQTT works at the same time. I expect HTTP request keep working nicely even if internet is not there. I tried it by disconnecting my wifi internet cable,keeping WiFi working but ESP was moving in a loop untill MQTT broker connected. I commented out this // while (!client.connected())   but now, it take almost 4-5 sec to response back the http request.,I want it to check for broker connection,try to connect,and come back fast to server HTTP request in the loop.
How can I achieve this?
Pl. advise.

**Comment by internetofhomethings** [comment]
Hi Santosh,
Thanks for your feedback. You have brought up some things that require clarification and will certainly help others not encounter the same pitfalls.
1. MQTT default packet size
   I forgot to mention in my post that the MQTT packet size in the PubSubClient library is limited to 128 bytes. Here is how to increase it so the MQTT reply will work with the example:

Open the file in your arduino libray folder ..Arduino/libraries/PubSubClient/src/PubSubClient.h

change:
// MQTT_MAX_PACKET_SIZE : Maximum packet size
#define MQTT_MAX_PACKET_SIZE 128

to:
// MQTT_MAX_PACKET_SIZE : Maximum packet size
#define MQTT_MAX_PACKET_SIZE 256

You might need to restart the Arduino IDE for this change to take effect.

Thanks again for your finding, I will update my post with this information.
1. "undefined" in reply

Please note the last statement in loop():

ReadSensors(2500);                   // Read 1 sensor every 2.5 seconds or longer

Only one sensor is read each 2.5 seconds. The Reply will be "undefined" until the first reading of a sensor.

The purpose for that is to keep the sensor reading from consuming all the CPU bandwidth. The negative consequences could be sluggish server response and possible watchdog timer timouts which would crash the CPU. You could try to speed that up by reducing 2500. It may work just fine, this is a very conservative design.
1. Internet connection down handler

This demo code is not robust enough to handle the case of a loss of connection gracefully. This might be an intrinsic constraint of the system and how the router resolves URLs, even local ones when the broadband connection is broken.  I am going to add some code to address this. Stay tuned for an answer to that issue soon.

**Comment by Santosh1971** [comment]
with this change everything is working fine as expected. Thanks

Few suggestion to make it more useful:-( it is my wishlist)
1. Internet connection down handler
2. Webconfig to Select the Home WiFi and the MQTT details as is done here
   http://www.esp8266.com/viewtopic.php?f=29&t=6127#p31994
3. The command should work in ESP8266 as AP mode also.( if no home wifi
   available)
4. Keeping the last state of "Devise"  in EEPROM and recalling it at power
   on.
5. FOTA ( Firmware over the air update of ESP8266 program.

Thanks agin
Regards,
Santosh

On Fri, Dec 4, 2015 at 8:49 PM, Dave notifications@github.com wrote:

> Hi Santosh,
> Thanks for your feedback. You have brought up some things that require
> clarification and will certainly help others not encounter the same
> pitfalls.
> 1. MQTT default packet size I forgot to mention in my post that the
>    MQTT packet size in the PubSubClient library is limited to 128 bytes. Here
>    is how to increase it so the MQTT reply will work with the example:
> 
> Open the file in your arduino libray folder
> ..Arduino/libraries/PubSubClient/src/PubSubClient.h
> 
> change:
> // MQTT_MAX_PACKET_SIZE : Maximum packet size
> #define MQTT_MAX_PACKET_SIZE 128
> 
> to:
> // MQTT_MAX_PACKET_SIZE : Maximum packet size
> #define MQTT_MAX_PACKET_SIZE 256
> 
> You might need to restart the Arduino IDE for this change to take effect.
> 
> Thanks again for your finding, I will update my post with this information.
> 1. "undefined" in reply
> 
> Please note the last statement in loop():
> 
> ReadSensors(2500); // Read 1 sensor every 2.5 seconds or longer
> 
> Only one sensor is read each 2.5 seconds. The Reply will be "undefined"
> until the first reading of a sensor.
> 
> The purpose for that is to keep the sensor reading from consuming all the
> CPU bandwidth. The negative consequences could be sluggish server response
> and possible watchdog timer timouts which would crash the CPU. You could
> try to speed that up by reducing 2500. It may work just fine, this is a
> very conservative design.
> 1. Internet connection down handler
> 
> This demo code is not robust enough to handle the case of a loss of
> connection gracefully. This might be an intrinsic constraint of the system
> and how the router resolves URLs, even local ones when the broadband
> connection is broken. I am going to add some code to address this. Stay
> tuned for an answer to that issue soon.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/internetofhomethings/ESP8266-MQTT-HTTP-Server/issues/1#issuecomment-161991844
> .

## 

Thanks and Regards,
Santosh Jha
8197239206

**Comment by internetofhomethings** [comment]
Thanks for your suggestions. They are all good ideas that would be useful for many applications. I'll be working on those items...look for a post with my implementation of these soon.

I did have a look at the configurable server (item #2) you referenced.When I attempted to compile the sketch, it became apparent that a different PubSubClient library was used than the one I had been using. After making some adjustments to match my library, the sketch compiled. It has a nice template to follow for configurable parameters and EEPROM read/write. The sketch is somewhat flakey, however as I have experienced several crashes. that is usually caused by a memory leak or a watchdog time-out. I have no desire to debug that code. But it does have some really useful components to reuse in your own code.

Thanks again for all your feedback,

Dave

PS. With your comment "with this change everything is working fine as expected" I am going to close this issue on GitHub.
